contracts/RadiCards.sol

```diff
@@ -3,234 +3,504 @@ pragma solidity ^0.4.24;
+    mapping(address => uint256) public ephemeralWalletCards; // ephemeral wallet => tokenId
+
+    struct RadiCard {
+        address gifter;
+        string message;
+        bool daiDonation;
+        uint256 giftAmount;
+        uint256 donationAmount;
+        Statuses status;
+        uint256 cardIndex;
+        uint256 benefactorIndex;
+    }
+
+    mapping(uint256 => Benefactor) public benefactors;
+    uint256[] internal benefactorsIndex;
+
+    mapping(uint256 => CardDesign) public cards;
+    uint256[] internal cardsIndex;
+
+    mapping(uint256 => RadiCard) public tokenIdToRadiCardIndex;
+
+    //total gifted/donated in ETH
+    uint256 public totalGiftedInWei;
+    uint256 public totalDonatedInWei;
+
+    //total gifted/donated in DAI
+    uint256 public totalGiftedInAtto; //SI prefix for 1x10-18 dai is Atto.
+    uint256 public totalDonatedInAtto;
+
+    event CardGifted(
+        address indexed _to,
+        uint256 indexed _benefactorIndex,
+        uint256 indexed _cardIndex,
+        address _from,
+        uint256 _tokenId,
+        Statuses status
     );
-  }
-
-  function tokenBenefactor(uint256 _tokenId)
-  public view
-  returns (
-    address _ethAddress,
-    string _name,
-    string _website,
-    string _logo
-  ) {
-    require(exists(_tokenId));
-    RadiCard memory _radiCard = tokenIdToRadiCardIndex[_tokenId];
-    Benefactor memory _benefactor = benefactors[_radiCard.benefactorIndex];
-    return (
-      _benefactor.ethAddress,
-      _benefactor.name,
-      _benefactor.website,
-      _benefactor.logo
-    );
-  }
-
-  function tokensOf(address _owner) public view returns (uint256[] _tokenIds) {
-    return ownedTokens[_owner];
-  }
-
-  function benefactorsKeys() public view returns (uint256[] _keys) {
-    return benefactorsIndex;
-  }
-
-  function cardsKeys() public view returns (uint256[] _keys) {
-    return cardsIndex;
-  }
-
-  function addBenefactor(uint256 _benefactorIndex, address _ethAddress, string _name, string _website, string _logo)
-  public onlyIfWhitelisted(msg.sender)
-  returns (bool) {
-    require(address(_ethAddress) != address(0), "Invalid address");
-    require(bytes(_name).length != 0, "Invalid name");
-    require(bytes(_website).length != 0, "Invalid name");
-    require(bytes(_logo).length != 0, "Invalid name");
-
-    benefactors[_benefactorIndex] = Benefactor(
-      _ethAddress,
-      _name,
-      _website,
-      _logo
-    );
-    benefactorsIndex.push(_benefactorIndex);
-
-    emit BenefactorAdded(_benefactorIndex);
-    return true;
-  }
 
-  function addCard(uint256 _cardIndex, string _tokenURI, bool _active)
-  public onlyIfWhitelisted(msg.sender)
-  returns (bool) {
-    require(bytes(_tokenURI).length != 0, "Invalid token URI");
-
-    cards[_cardIndex] = CardDesign(
-      _tokenURI,
-      _active
+    event LogClaim(
+            address indexed ephemeralAddress,
+            address indexed sender,
+            uint tokenId,
```

Comment:
Maybe add `indexed` to the `tokenId` as the final and 3rd index on this event

Reply:
done